File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root, int height, map<int, vector<int>>& mp) {
        if (!root)
            return;
        mp[height].push_back(root->val);
        if (root->left)
            helper(root->left, height + 1, mp);
        if (root->right)
            helper(root->right, height + 1, mp);
    }

    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans;
        map<int, vector<int>> mp;
        helper(root, 0, mp);
        for (auto i : mp)
            ans.push_back(i.second.back());
        return ans;
    }
};

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null)
            return ans;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            int size = q.size();
            int lastVal = 0;
            while (size != 0) {
                TreeNode top = q.poll();
                lastVal = top.val;
                if (top.left != null)
                    q.add(top.left);
                if (top.right != null)
                    q.add(top.right);
                size--;
            }
            ans.add(lastVal);
        }
        return ans;
    }
}